_posts/2023-11-03-visual-testing-using-playwright.md

```diff
@@ -0,0 +1,201 @@
+
+
+## **Why is visual testing important?**
+
+- Visual testing captures defects that other types of UI tests cannot
+- Visual testing is essential for ensuring a positive user experience with your app
+- Maintaining a visually perfect UI is important for retaining customers
+- Visual testing is important to reduce the revenue impact
+
+**List of Visual Bugs:-** below are the list of visual defects that may occur
+- Pixel-by-pixel differences
+- Alignment shifts
+- Page layout
+- Rendering problems
+- Element overlap
+- Responsive layout
+- Font differences
+- Color differences
+
+**Example 1:- Element Overlap Scenario**
+
+{% include
+  components/figure.html
+  url="/assets/images/post/visual-testing1.png"
+%}
+
+Look at this UI, how did this happen in production?
+The text and advertisement are crammed together. If this was your advertisement, do you think there would be a revenue impact? Absolutely
+
+These aren’t cosmetic issues. In each case, **visual bugs are blocking revenue**
+
+Why do these visual bugs occur? Don’t they do functional testing? They do, but it’s not enough
+
+Many believe that functional tests are enough to catch visual bugs, but that is not true. Functional tests validate only the functional behavior of the app
+
+Visual bugs are rendering issues. And rendering validation is not what functional testing tools are designed to catch.
+
+**Example 2:- Different visual bugs on each case**
+
+{% include
+  components/figure.html
+  url="/assets/images/post/visual-testing2.png"
+%}
+
+In the preceding example each snapshot has rendered with different visual error
+
+## **How can Visual Testing help?**
+
+
+{% include
+  components/figure.html
+  url="/assets/images/post/visual-testing3.png"
+  description="Visual aspects of the quality"
+%}
+
+**Benefits of Visual Testing**
+
+- Provide Confidence for visual aspects of the product that can’t be covered by lower level validations (Even by E2E)
+- Easy to cover cases like RTL changes or Theming if already have original tests
+- For some functional tests it is **faster to develop** Visual tests than functional
+- **Maintenance of the visual tests can be cheaper than functional tests** (All you need is just update baselines, you have no dependencies to test ids and html structure)
+<br>
+<br>
+
+## **Why can’t functional tests cover visual issues?**
+
+Functional test scripts can validate the size, position, and color scheme of visual elements. But if we follow this approach, then our test scripts will soon balloon in size due to checkpoint bloat
+
+let’s look at an Instagram advertisement screen that’s properly rendered
+
+{% include
+  components/figure.html
+  url="/assets/images/post/visual-testing4.png"
+  description="Advertisement screen that has properly rendered"
+%}
+
+If we look at the advertisement, there are 21 visual elements like, various icons, text.(This ignores iOS elements at the top like Wi-Fi signal and time, since those aren’t controlled by the Instagram app)
```

Comment:
```suggestion
If we look at the advertisement, there are 21 visual elements, like various icons, text – this ignores iOS elements at the top like Wi-Fi signal and time, since those aren’t controlled by the Instagram app.
```